.github/workflows/ci-cd.yml

```diff
@@ -0,0 +1,236 @@
+        uses: actions/checkout@v2
+
+      # Tooling setup
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Install graphviz
+        run: |
+          sudo apt-get update -y
+          sudo apt-get install -y graphviz
+
+      - name: Install Python tools
+        run: |
+          python3 -m pip install --upgrade pip
+          pip install flake8 wheel
+
+      # Information setup
+
+      - if: matrix.python-version == '3.9'
+        name: Pipeline data gatherer
+        id: data_gatherer
+        run: |
+          # Get default branch
+          REPO="${{ github.repository }}"
+          DEFAULT_BRANCH=`(curl -X GET https://api.github.com/repos/$REPO) | jq '.default_branch'`
+
+          echo "::set-output name=default_branch::$(echo $DEFAULT_BRANCH)"
+
+      - if: matrix.python-version == '3.9'
+        name: Pipeline conditionals handler
+        id: conditionals_handler
+        run: |
+          DEFAULT_BRANCH="${{ steps.data_gatherer.outputs.default_branch }}"
+          GITHUB_REF="${{ github.ref }}"
+          CURRENT_BRANCH=`echo $GITHUB_REF | sed -e "s/^refs\/heads\///"`
+          GITHUB_EVENT_NAME="${{ github.event_name }}"
+          IS_DEFAULT_BRANCH='false'
+          IS_PUSH='false'
+          IS_PUSH_TO_DEFAULT_BRANCH='false'
+
+          if [ $CURRENT_BRANCH == $DEFAULT_BRANCH ]; then
+            IS_DEFAULT_BRANCH='true'
+          fi
+          if [ $GITHUB_EVENT_NAME == 'push' ]; then
+            IS_PUSH='true'
+          fi
+          if [ $CURRENT_BRANCH == $DEFAULT_BRANCH ] && [ $GITHUB_EVENT_NAME == 'push' ]; then
+            IS_PUSH_TO_DEFAULT_BRANCH='true'
+          fi
+
+          echo "::set-output name=is_default_branch::$(echo $IS_DEFAULT_BRANCH)"
+          echo "::set-output name=is_push::$(echo $IS_PUSH)"
+          echo "::set-output name=is_push_to_default_branch::$(echo $IS_PUSH_TO_DEFAULT_BRANCH)"
+
+      - if: steps.conditionals_handler.outputs.is_push_to_default_branch == 'true' && matrix.python-version == '3.9'
+        name: Project version/deploy determiner
+        id: proj_ver_determiner
+        run: |
+          git fetch --all --tags
+
+          PACKAGE_VERSION=$(cat setup.py \
+            | grep version \
+            | head -1 \
+            | sed "s/[version=\', ]//g")
+          PACKAGE_VERSION_TAG=`echo v$PACKAGE_VERSION | sed -e 's/[[:space:]]//'`
+          LATEST_GITHUB_TAG=`echo $(git tag | sort --version-sort | tail -n1)`
+          SHOULD_DEPLOY="false"
+
+          # download semver compare tool
+          curl https://raw.githubusercontent.com/Ariel-Rodriguez/sh-semversion-2/main/semver2.sh -o semver2.sh
+          chmod +x semver2.sh
+
+          if [ -z "$LATEST_GITHUB_TAG" ]; then
+            SHOULD_DEPLOY="true"
+          else
+            if [ `echo $(./semver2.sh $PACKAGE_VERSION_TAG $LATEST_GITHUB_TAG)` = 1 ]; then
```

Comment:
This is a really cool implementation, but I see a potential small issue that both `p(a[1])>p(a[2])` and `p(a[1])<p(a[2])` return the same value `0`, which means if by accident a lower semver is committed it would break the pipeline at the deployment stage, since such version already exist on GitHub and PyPI and would create unnecessary hassle.

What do you think ?

Reply:
@aleks-ivanov I'm all for robust pipeline code ;-)

I guess I didn't read the semver2.sh code well enough, I see now that it is a cmp (returns -1, 0, 1).
Since Python lost cmp in py3, the way to do it would be (https://docs.python.org/3.0/whatsnew/3.0.html#ordering-comparisons):
```
import sys, packaging.version as v
p = v.parse
s = sys.argv
a = p(s[1])
b = p(s[2])
print((a > b) - (a < b))
```
which is slightly longer but returns -1, 0, and 1 as expected:
```
python -c "import sys,packaging.version as v;p=v.parse;s=sys.argv;a=p(s[1]);b=p(s[2]);print((a>b)-(a<b))" 1.9.14 1.9.14
```

Because of the codecov bash uploader issue earlier this year we're very uncomfortable with downloading and running bash scripts (I suppose we could check the sha...).